src/typing.py

```diff
@@ -1916,7 +1916,21 @@ def __new__(cls, typename, bases, ns):
             raise TypeError("Class syntax for NamedTuple is only supported"
                             " in Python 3.6+")
         types = ns.get('__annotations__', {})
-        return _make_nmtuple(typename, types.items())
+        nm_tpl = _make_nmtuple(typename, types.items())
+        saw_default = False
+        defaults = []
+        defaults_dict = {}
+        for field_name in types:
+            if field_name in ns:
+                saw_default = True
+                default_value = ns[field_name]
+                defaults.append(default_value)
+                defaults_dict[field_name] = default_value
+            elif saw_default:
+                raise TypeError('Non-default namedtuple field cannot follow default field')
+        nm_tpl.__new__.__defaults__ = tuple(defaults)
```

Comment:
As I understand it, dict ordering is an implementation detail of CPython 3.6 and a conforming Python 3.6 implementation could choose to not implement ordering, so this code shouldn't rely on dict ordering (except for class namespaces and **kwargs because of PEP 520 and PEP 468).